src/main/java/com/commercetools/project/sync/SyncerFactory.java

```diff
@@ -55,122 +61,172 @@ public static SyncerFactory of(
+                      format(
+                          "Unknown argument \"%s\" supplied to \"-%s\" or \"--%s\" option! %s",
+                          trimmedValue,
+                          SYNC_MODULE_OPTION_SHORT,
+                          SYNC_MODULE_OPTION_LONG,
+                          SYNC_MODULE_OPTION_DESCRIPTION);
+                  throw new IllegalArgumentException(errorMessage);
+                }
+              })
+          .collect(Collectors.toList());
     }
+  }
 
-    Syncer<
-            ? extends ResourceView,
-            ?,
-            ? extends BaseSyncStatistics,
-            ? extends BaseSyncOptions<?, ?>,
-            ? extends QueryDsl<?, ?>,
-            ? extends BaseSync<?, ?, ?>>
-        syncer;
+  private static Collection<List<SyncModuleOption>> groupSyncModuleOptions(
+      List<SyncModuleOption> syncModuleOptions) {
+    Map<SyncModuleOption, Integer> syncModuleOptionWithReferencesCount =
+        syncModuleOptions
+            .stream()
+            .collect(
+                Collectors.toMap(
+                    key -> key,
+                    option ->
+                        (option.getReferences().isEmpty())
+                            ? 0
+                            : countReferences(option.getReferences(), syncModuleOptions)));
 
-    try {
-      syncer = buildSyncer(syncOptionValue, runnerNameOptionValue, isSyncProjectSyncCustomObjects);
-    } catch (IllegalArgumentException exception) {
+    Map<Integer, List<SyncModuleOption>> groupedSyncModuleOptions = new TreeMap<>();
+    for (Map.Entry<SyncModuleOption, Integer> syncOptionWithReferenceCount :
+        syncModuleOptionWithReferencesCount.entrySet()) {
+      if (!groupedSyncModuleOptions.containsKey(syncOptionWithReferenceCount.getValue())) {
+        List<SyncModuleOption> syncOptionsList = new ArrayList<>();
+        for (Map.Entry<SyncModuleOption, Integer> syncModuleOptionIntegerEntry :
+            syncModuleOptionWithReferencesCount.entrySet()) {
+          if (0
+              == syncOptionWithReferenceCount
+                  .getValue()
+                  .compareTo(syncModuleOptionIntegerEntry.getValue())) {
+            syncOptionsList.add(syncModuleOptionIntegerEntry.getKey());
+          }
+        }
+        groupedSyncModuleOptions.put(syncOptionWithReferenceCount.getValue(), syncOptionsList);
+      }
+    }
+    return groupedSyncModuleOptions.values();
+  }
 
-      return exceptionallyCompletedFuture(exception);
+  private static int countReferences(
+      List<SyncModuleOption> syncOptionDependencies, List<SyncModuleOption> syncModuleOptions) {
+    if (null == syncOptionDependencies || syncOptionDependencies.isEmpty()) {
+      return 1;
+    }
+    int count = 0;
+    for (SyncModuleOption value : syncOptionDependencies) {
+      if (!syncModuleOptions.contains(value)) {
+        continue;
+      }
+      count += countReferences(value.getReferences(), syncModuleOptions);
```

Comment:
Yes, Here we are trying to keep the same way how we documented. The resources which has references for other resources are added to the SyncModuleOption. Initially we named it is as Dependencies(Doesn't look good). You can even suggest a better name for this(If you have any).

Reply:
Hmm, Jude seems right, It's not a direct reference in commercetools resource, because the option is not a commercetools resource, so here it's more like the relations between options and a grouping for the order of execution. So `referencedOptions` might be.. it's hard to name though because it's needed in the grouping algorithm 🤷🏽

Reply:
referencedOptions or referencedModuleOptions?

Reply:
I might avoid using the word 'reference' for this field. Rather the word `essentialSyncOptions` would imply there is a prerequisite to executing a sync operation.

Reply:
https://github.com/commercetools/commercetools-project-sync/pull/219/commits/b0015c3ed6f5d814a64a95e1df6b5ba7e8acc7ab